docs/configuration.rst
Configure the Bundle
====================

The bundle allows you to specify different Message Queue providers - however,
Amazon AWS and IronMQ are the only ones currently supported. Blocking, synchronous queues
are also supported through the ``sync`` driver to aid development and debugging.

We are actively looking to add more and would be more than happy to accept contributions.

Providers
---------

This bundle allows you to configure and use multiple supported providers with in the same
application. Each queue that you create is attached to one of your registered providers
and can have its own configuration options.

Providers may have their own dependencies that should be added to your ``composer.json`` file.

For specific instructions on how to configure each provider, please view their documents.

.. toctree::
    :maxdepth: 2

    aws-provider
    iron-mq-provider
    sync-provider
    file-provider
    custom-provider

Caching
-------

Providers can leverage a caching layer to limit the amount of calls to the Message Queue
for basic lookup functionality - this is important for things like AWS's ARN values, etc.

By default the library will attempt to use file cache, however you can pass your
own cache service, as long as its an instance of ``Doctrine\Common\Cache\Cache``.

The configuration parameter ``cache_service`` expects the container service id of a registered
Cache service. See below.

.. code-block:: yaml

    #app/config.yml

    services:
        my_cache_service:
            class: My\Caching\CacheService

    uecode_qpush:
        cache_service: my_cache_service

**Note:** *Though the Queue Providers will attempt to create queues if they do not exist when publishing or receiving messages,
it is highly recommended that you run the included console command to build queues and warm cache from the CLI beforehand.*

Queue Options
-------------

Each queue can have their own options that determine how messages are published or received.
The options and their descriptions are listed below.

+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| Option                          | Description                                                                                | Default Value |
+=================================+============================================================================================+===============+
| ``queue_name``                  | The name used to describe the queue on the Provider's side                                 | ``null``      |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``push_notifications``          | Whether or not to POST notifications to subscribers of a Queue                             | ``false``     |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``push_notifications_only``     | AWS Only for now. Disable the Queue Polling if you do not require message for 1 hour+      | ``false``     |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``notification_retries``        | How many attempts notifications are resent in case of errors - if supported                | ``3``         |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``message_delay``               | Time in seconds before a published Message is available to be read in a Queue              | ``0``         |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``message_timeout``             | Time in seconds a worker has to delete a Message before it is available to other workers   | ``30``        |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``message_expiration``          | Time in seconds that Messages may remain in the Queue before being removed                 | ``604800``    |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``messages_to_receive``         | Maximum amount of messages that can be received when polling the queue                     | ``1``         |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``receive_wait_time``           | If supported, time in seconds to leave the polling request open - for long polling         | ``3``         |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``fifo``                        | If supported (only aws), sets queue into FIFO mode                                         | ``false``     |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``content_based_deduplication`` | If supported (only aws), turns on automatic deduplication id based on the message content  | ``false``     |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+
| ``subscribers``                 | An array of Subscribers, containing an ``endpoint`` and ``protocol``                       | ``empty``     |
+---------------------------------+--------------------------------------------------------------------------------------------+---------------+

Symfony Application as a Subscriber
-----------------------------------

The QPush Bundle uses a Request Listener which will capture and dispatch notifications from your queue providers for you. The specific route you use does not matter.

In most cases, it is recommended to just list the host or domain for your Symfony application as the ``endpoint`` of your subscriber.  You do not need to create a new action for QPush to receive messages.

Logging with Monolog
--------------------

By default, logging is enabled in the Qpush Bundle and uses Monolog, configured
via the MonologBundle. You can toggle the logging behavior by setting
``logging_enabled`` to ``false``.

Logs will output to your default Symfony environment logs using the 'qpush' channel.

Example Configuration
---------------------

A working configuration would look like the following

.. code-block:: yaml

    uecode_qpush:
        cache_service: null
        logging_enabled: true
        providers:
            aws:
                driver: aws #optional for providers named 'aws' or 'ironmq'
                key: YOUR_AWS_KEY_HERE
                secret: YOUR_AWS_SECRET_HERE
                region: YOUR_AWS_REGION_HERE
            another_aws_provider:
                driver: aws #required for named providers
                key: YOUR_AWS_KEY_HERE
                secret: YOUR_AWS_SECRET_HERE
                region: YOUR_AWS_REGION_HERE
            ironmq:
                driver: aws #optional for providers named 'aws' or 'ironmq'
                token: YOUR_IRONMQ_TOKEN_HERE
                project_id: YOUR_IRONMQ_PROJECT_ID_HERE
            in_band:
                driver: sync
            custom_provider:
                driver: custom
                service: YOUR_CUSTOM_SERVICE_ID
        queues:
            my_queue_key:
                provider: ironmq #or aws or in_band or another_aws_provider
                options:
                    queue_name:                  my_actual_queue_name
                    push_notifications:          true
                    push_notifications_only:     false
                    notification_retries:        3
                    message_delay:               0
                    message_timeout:             30
                    message_expiration:          604800
                    messages_to_receive:         1
                    receive_wait_time:           3
                    fifo:                        false
                    content_based_deduplication: false
                    subscribers:
                        - { endpoint: http://example1.com/, protocol: http }
                        - { endpoint: http://example2.com/, protocol: http }
            my_fifo_queue_key:
                provider: aws
                options:
                    queue_name:                  my_actual_queue_name.fifo
                    push_notifications:          false
                    notification_retries:        3
                    push_notifications_only:     false
                    message_delay:               0
                    message_timeout:             30
                    message_expiration:          604800
                    messages_to_receive:         1
                    receive_wait_time:           3
                    fifo:                        true
                    content_based_deduplication: true
                    subscribers:
                        - { endpoint: http://example1.com/, protocol: http }
                        - { endpoint: http://example2.com/, protocol: http }
                        
Note that `FIFO` queues are not currently compatible with `push_notifications`. For more information, see: http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-subscribe-queue-sns-topic.html


docs/_build/html/_sources/configuration.txt
Configure the Bundle
====================

The bundle allows you to specify different Message Queue providers - however, 
Amazon AWS and IronMQ are the only ones currently supported. 

We are actively looking to add more and would be more than happy to accept contributions.

Providers
---------

This bundle allows you to configure and use multiple supported providers with in the same 
application. Each queue that you create is attached to one of your registered providers
and can have its own configuration options.

Providers may have their own dependencies that should be added to your ``composer.json`` file.

For specific instructions on how to configure each provider, please view their documents.

.. toctree::
    :maxdepth: 1

    aws-provider
    iron-mq-provider

Caching
-------

Providers can leverage a caching layer to limit the amount of calls to the Message Queue
for basic lookup functionality for things like the Queue ARN, etc.

By default the library will attempt to use file cache, however you can pass your
own cache service, as long as its an instance of ``Doctrine\Common\Cache\Cache``.

The configuration parameter ``cache_service`` expects the container service id of a registered
Cache service. See below.

.. code-block:: yaml

    #app/config.yml

    services:
        my_cache_service:
            class: My\Caching\CacheService

    uecode_qpush:
        cache_service: my_cache_service

**Note:** *Though the Queue Providers will attempt to create queues if they do not exist when publishing or receiving messages,
it is highly recommended that you run the included console command to build queues and warm cache from the CLI before hand.*

Queue Options
-------------

Each queue can their have own options that determine how messages are published or receieved. 
The options and their descriptions are listed below.

+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| Option                   | Description                                                                               | Default Value |
+==========================+===========================================================================================+===============+
| ``push_notifications``   | Whether or not to POST notifications to subscribers of a Queue                            | ``false``     |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``notification_retries`` | How many attempts notifications are resent in case of errors - if supported               | ``3``         |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``message_delay``        | Time in seconds before a published Message is available to be read in a Queue             | ``0``         |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``message_timeout``      | Time in seconds a worker has to delete a Message before its available to other workers    | ``30``        |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``message_expiration``   | Time in seconds that Messages may remain in the Queue before being removed                | ``604800``    |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``messages_to_receive``  | Maximum amount of messages that can be received when polling the queue                    | ``1``         |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``receive_wait_time``    | If supported, time in seconds to leave the polling request open - for long polling        | ``3``         |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``subscribers``          | An array of Subscribers, containing an ``endpoint`` and ``protocol``                      | ``empty``     |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+

Example Configuration
---------------------

However, a working configuration would look like the following

.. code-block:: yaml

    uecode_qpush:
        cache_service: null
        logging_enabled: true
        providers:
            aws:
                key:
                secret:
                region:
            ironmq:
                token:
                project_id:
        queues:
            default:
                provider: aws #or ironmq
                options:
                    push_notifications:     true
                    notification_retries:   3
                    message_delay:          0
                    message_timeout:        30
                    message_expiration:     604800
                    messages_to_receive:    1
                    receive_wait_time:      3
                    subscribers:
                        - { endpoint: http://example1.com/qpush, protocol: http }
                        - { endpoint: http://example2.com/qpush, protocol: http }




CHANGELOG.md
03-01-2014

* Notification now extends Message
* ProviderInterface renamed to Provider

src/DependencyInjection/Configuration.php
<?php

/**
 * Copyright 2014 Underground Elephant
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @package     qpush-bundle
 * @copyright   Underground Elephant 2014
 * @license     Apache License, Version 2.0
 */

namespace Uecode\Bundle\QPushBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * @author Keith Kirk <kkirk@undergroundelephant.com>
 */
class Configuration implements ConfigurationInterface
{
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode    = $treeBuilder->root('uecode_qpush');

        $rootNode
            ->addDefaultsIfNotSet()
            ->children()
                ->scalarNode('cache_service')
                    ->defaultNull()
                ->end()
                ->booleanNode('logging_enabled')
                    ->defaultTrue()
                ->end()
                ->append($this->getProvidersNode())
                ->append($this->getQueuesNode())
            ->end()
        ;

        return $treeBuilder;
    }

    private function getProvidersNode()
    {
        $treeBuilder    = new TreeBuilder();
        $node           = $treeBuilder->root('providers');
        $requirements   = [
            'aws' => [],
            'ironmq' => ['token', 'project_id'],
            'sync' => [],
            'custom' => ['service'],
            'file' => ['path'],
            'doctrine' => []
        ];

        $node
            ->useAttributeAsKey('name')
            ->prototype('array')
                ->treatNullLike([])
                ->children()
                    ->enumNode('driver')
                        ->isRequired()
                        ->values(array_keys($requirements))
                    ->end()
                    // IronMQ
                    ->scalarNode('token')->end()
                    ->scalarNode('project_id')->end()
                    ->scalarNode('service')->end()
                    ->enumNode('host')
                        ->defaultValue('mq-aws-eu-west-1-1')
                        ->values([
                            'mq-aws-eu-west-1-1',
                            'mq-aws-us-east-1-1',
                        ])
                    ->end()
                    ->scalarNode('port')
                        ->defaultValue('443')
                    ->end()
                    ->scalarNode('api_version')
                        ->defaultValue(3)
                    ->end()
                    // AWS
                    ->scalarNode('key')->end()
                    ->scalarNode('secret')->end()
                    ->scalarNode('region')
                        ->defaultValue('us-east-1')
                    ->end()
                    // File
                    ->scalarNode('path')->end()
                   // Doctrine
                    ->scalarNode('entity_manager')
                         ->defaultValue('doctrine.orm.default_entity_manager')
                    ->end()
                ->end()

                ->validate()
                ->always()
                ->then(function (array $provider) use ($node, $requirements) {
                    foreach ($requirements[$provider['driver']] as $requirement) {
                        if (empty($provider[$requirement])) {
                            throw new \InvalidArgumentException(
                                sprintf('%s queue providers must have a %s; none provided', $provider['driver'], $requirement)
                            );
                        }
                    }

                    return $provider;
                })
            ->end()
        ;

        return $node;
    }

    private function getQueuesNode()
    {
        $treeBuilder = new TreeBuilder();
        $node        = $treeBuilder->root('queues');

        $node
            ->requiresAtLeastOneElement()
            ->useAttributeAsKey('name')
            ->prototype('array')
                ->children()
                    ->scalarNode('provider')
                        ->isRequired()
                        ->info('The Queue Provider to use')
                    ->end()
                    ->arrayNode('options')
                        ->children()
                            ->scalarNode('queue_name')
                                ->defaultNull()
                                ->info('The actual name of the queue')
                            ->end()
                            ->booleanNode('push_notifications')
                                ->defaultFalse()
                                ->info('Whether notifications are sent to the subscribers')
                            ->end()
                            ->booleanNode('push_notifications_only')
                                ->defaultFalse()
                                ->info('Where to disable queue polling')
                            ->end()
                            ->scalarNode('push_type')
                                ->defaultValue('multicast')
                                ->info('Whether the push queue is multicast or unicast')
                                ->example('unicast')
                            ->end()
                            ->scalarNode('notification_retries')
                                ->defaultValue(3)
                                ->info('How many attempts the Push Notifications are retried if the Subscriber returns an error')
                                ->example(3)
                            ->end()
                            ->scalarNode('notification_retries_delay')
                                ->defaultValue(60)
                                ->info('Delay between each Push Notification retry in seconds')
                                ->example(3)
                                ->end()
                            ->scalarNode('message_delay')
                                ->defaultValue(0)
                                ->info('How many seconds before messages are inititally visible in the Queue')
                                ->example(0)
                            ->end()
                            ->scalarNode('message_timeout')
                                ->defaultValue(30)
                                ->info('How many seconds the Queue hides a message while its being processed')
                                ->example(30)
                            ->end()
                            ->scalarNode('message_expiration')
                                ->defaultValue(604800)
                                ->info('How many seconds a message is kept in Queue, the default is 7 days (604800 seconds)')
                                ->example(604800)
                            ->end()
                            ->scalarNode('messages_to_receive')
                                ->defaultValue(1)
                                ->info('Max amount of messages to receive at once - an event will be fired for each individually')
                                ->example(1)
                            ->end()
                            ->scalarNode('receive_wait_time')
                                ->defaultValue(0)
                                ->info('How many seconds to Long Poll when requesting messages - if supported')
                                ->example(3)
                            ->end()
                            ->scalarNode('rate_limit')
                                ->defaultValue(-1)
                                ->info('How many push requests per second will be triggered. -1 for unlimited, 0 disables push')
                                ->example(1)
                            ->end()
                            ->booleanNode('fifo')
                                ->defaultFalse()
                                ->info('If the queue is FIFO (aws)')
                            ->end()
                            ->booleanNode('content_based_deduplication')
                                ->defaultFalse()
                                ->info('If the FIFO queue has content based deduplication (aws)')
                            ->end()
                            ->append($this->getSubscribersNode())
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;

        return $node;
    }

    private function getSubscribersNode()
    {
        $treeBuilder = new TreeBuilder();
        $node        = $treeBuilder->root('subscribers');

        $node
            ->prototype('array')
                ->children()
                    ->scalarNode('endpoint')
                        ->info('The url or email address to notify')
                        ->example('http://foo.bar/qpush/')
                    ->end()
                    ->enumNode('protocol')
                        ->values(['email', 'http', 'https'])
                        ->info('The endpoint type')
                        ->example('http')
                    ->end()
                ->end()
            ->end()
        ;

        return $node;
    }
}
